### ISSUE TITLE
Add endpoint for getting list of components with their versions

### ISSUE THREAD

**Issue by ferewuz** [opentable/oc#324]
Following up on discussion in gitter channel, to enable clients to always make versioned requests to oc-registry it would be useful having a separate endpoint returning all components with their versions included.

Implementation would be similar to `list-components` when called without `text\html` headers (https://github.com/opentable/oc/blob/master/src/registry/routes/list-components.js)

My thinking was adding a route called `list-components-versions`.
Return value will be a json, containing components and sorted list of versions, something like:
```
"components": {
    "http://oc-registry.com/component1" : [ '1.0.0', '1.1.0', '2.0.0'],
    "http://oc-registry.com/component2" : [ '1.0.0', '2.1.0', '3.0.0']
}
```

Possible extension would be to limit the components you want to get the version for, using query params. Not a must-have for first version.

I am happy to go forward with implementation and PR if that makes sense to you @matteofigus and other contributors?

**Comment by matteofigus** [comment]
Hi @ferewuz, thanks for the issue. I spent some time to think on this, and I actually may have another idea.

The goal of this feature is to enable clients to understand what matches a requested version and understand if something changed given an amount of time. In theory, you can already do this, but I would like your opinion on that.

This is an example for the batch endpoint request, already implemented:

```js
POST https://oc-registry.com
{
  components: [{
    name: "component1",
    version: "1.x.x"
  }, {
    name: "component2",
    version: "1.4.x",
    parameters: { a: 12354 }
  }]
}

//would give you back something like:

{
  components: [{
    status: 200,
    response: {
      name: "component1",
      href: "https://oc-registry.com/component1",
      html: "...",
      requestVersion: "1.x.x",
      version: "1.2.3"
    }
  }, {
    status: 200,
    response: {
      name: "component2",
      href: "https://oc-registry.com/component2?a=12345
      html: "...",
      requestVersion: "1.4.x",
      version: "1.4.3"
    }
  }]
}
```

The good things about using the existing batch endpoint to achieve the same result:
1) You can ask only for the components you care about
2) You delegate to the registry the version resolving (with the other approach you may need to pull the `semver` module into the clients to handle to currently supported semver scenarios and perform the matching on the client side) - so clients can just compare if resolved versions are the same compared to a previous request

Bad things:
1) You get more than you want, like the component response, etc.
2) If a component has mandatory parameters etc, you actually need to include in the payload those parameters for of a proper request

So my idea is, what about modifying the existing BATCH endpoint to do what we want here.
For example sending a specific Accept, like `application/vnd.oc.versioninfo+json` it would just require an array of components and versions, and respond only with name, requestVersion, and version?

The cool thing about this is that the request and response would be super small - good if you want to implement like a short cache like 30s. Also this would work for a registry like ours that has a combination of 2500 components/versions.

What do you think?

**Comment by ivan-dimitrov-skyscanner** [comment]
Hello @matteofigus, @ferewuz,

May I jump in the conversation...

@matteofigus, I like your idea in general but may I shift the proposed endpoint from not being POST and not being the batch request.
We already have the information for all the latest versions of all the components present on the default registry page home "/". However it is in an HTML format. What if we modify this request to anticipate the Accept HTTP request header and if it is:
Accept: application/json
then we could return the information from the home page rendered in JSON?

Then further we could mirror this endpoint to /list or /list_info if necessary.

Regards,
Ivan

**Comment by ferewuz** [comment]
Hey @matteofigus,

Thanks for looking into it, and looking through it I agree that actually does look cleaner and would make calls much smaller and more concise!

In this case I would definitely go for adding the part you mentioned at the bottom

> sending a specific Accept, like application/vnd.oc.versioninfo+json

otherwise slow components could have a big impact on response time and we could be making lots of unneeded underlying requests and also the point of adding unneeded "dummy" parameters.

@ivan-dimitrov-skyscanner 
I think in this case POST request still makes sense, as we are sending the information about which components we want. 
And going with the implementation formulated by @matteofigus we wouldn't need the `/list` call at all. 
Basically all the calls `oc-client` would be making would still be to the `POST https://oc-registry.com`

Only difference being, they would sometimes request just the version, when polling for version updates and sometimes making a full requests with parameters included when requesting the full component responses (rendered or un-rendered).

Definitely a +1 from me for that implementation

**Comment by matteofigus** [comment]
@ivan-dimitrov-skyscanner hi! Wanted to let you know that if you GET both the registry home and the /component-name/~info with `Accept: application/json` you can already see everything in JSON format, including all the versions (semver sorted) of each component. But I think we need to implement a single endpoint to make everything a bit more performant.

@ferewuz glad you like it. Just to summarise, the idea would be that for a req like:
```js
POST https://oc-registry.com -H "Accept: application/vnd.oc.info+json"
{
  components: [{
    name: 'component1',
    version: '1.X.X'
  }, {
    name: 'component2',
    version: '~1.2.4'
  }]
}
```
would respond like
```js
{
  components: [{
    status: 200,
    response: {
      name: 'component1',
      version: '1.2.4',
      requestVersion: '1.X.X'
    }
  }, {
    status: 404,
    response: {
      name: 'component2',
      requestVersion: '~1.2.4',
      error: 'Error detail...something like component with requested version not found
   }]
}
```
Specifically, I like the idea of still keeping response/status because when we'll talk about caching, we will probably be able to add here something component-based like `{ headers: { 'cache-control': 'max-age=120'}}`. So this route will contain both versions and cache policies, potentially. Only problem of that is it wouldn't be an endpoint about versions only. So probably something like `Accept: application/vnd.oc.info+json` may be more appropriate? Allows us more flexibility when we'll add more stuff later on. What do you think?

**Comment by ferewuz** [comment]
@matteofigus 

Agree with the idea and request examples, looks good to me!

I do have some  about cache-headers, I don't think it would make sense extend the response with cache headers here?

My thinking is that the cache-headers (and any other headers) are part of the actual component request later (batch or not). 
`oc-client`/browser/CDN can then cache appropriately based on the first response coming back from the `oc-registry`. 

Separating the cache headers to response only achieves 2 things:

- Caching behaviour is uniform across environments, every consumer depends on cache-headers as part of the component response
- It allows a possibility of components having dynamic cache-headers in the future. 
Static cache-headers per component might work in most cases, but we have examples in our current web architecture setup where component will sometimes return a partial response, because of underlying dependency problems or similar.
In that case the component would set headers as `no-cache` or something much lower than usual, making the consumer retry until it gets a full version and only store that one in cache for future use

I would still agree that `Accept: application/vnd.oc.info+json ` makes more sense in any case, as we may want to extend the response with other info at some point and it contains a bit more that just versions already as is. But I would think that headers should not be part of the info response here.

What do you think?

**Comment by matteofigus** [comment]
Oh, yes, good point actually about making dynamic cache headers, I agree.

Ok so, I think we can proceed with this as is and then bring the cache topic back to a separate issue later on. 

Are you happy to try working on this? Let me know if you have any more questions

**Comment by ferewuz** [comment]
Ok!

I (or someone else from our team) will get on with implementation and try to get a PR ready soon.

If we have any more questions or issues we will report back here :)

**Comment by ivan-dimitrov-skyscanner** [comment]
Hi @matteofigus, I'm starting the work on this...
It will include a modification of the registry so that it will anticipate the proposed header (Accept: application/vnd.oc.info+json) and a method in the client (GetComponentsInfo) to leverage this new functionality.
